## Load Libraries

In [2]:
import pandas as pd
import numpy as np

from sklearn.model_selection import train_test_split
from sklearn.metrics import r2_score

from sklearn.linear_model import LinearRegression
from sklearn.preprocessing import PolynomialFeatures
from sklearn.ensemble import RandomForestRegressor, GradientBoostingRegressor
from sklearn.metrics import mean_squared_error

## Import cleaned csv 

In [3]:
df = pd.read_csv("cleaned_housing.csv")
df.head()

,price,bedrooms,bathrooms,livingArea,lotSize,yearBuilt,propertyTaxRate,taxAssessedValue,zestimate,rentZestimate,latitude,longitude,resoFacts/stories,resoFacts/hasGarage,resoFacts/hasHeating,resoFacts/hasCooling,resoFacts/hasFireplace,resoFacts/isNewConstruction,homeType
0,315400,3.0,3.0,1806.0,1555.0,2007.0,1.12,273960.0,317406.0,1950.0,45.543571,-122.441765,2.0,0,1,0,1,0,TOWNHOUSE
1,400000,3.0,3.0,1518.0,3484.0,2001.0,1.12,426330.0,573700.0,2545.0,45.547585,-122.453224,2.0,0,1,1,1,0,SINGLE_FAMILY
2,512000,4.0,3.0,2724.0,9583.0,1982.0,1.12,397560.0,512355.0,1995.0,45.488228,-122.444420,2.0,0,1,1,1,0,SINGLE_FAMILY
3,348000,3.0,1.0,1217.0,13939.0,1953.0,1.12,269520.0,573700.0,1695.0,45.526634,-122.464088,1.0,0,1,0,1,0,SINGLE_FAMILY
4,510000,6.0,3.0,2400.0,8545.0,1967.0,1.12,421980.0,573700.0,1395.0,45.511238,-122.431534,2.0,0,1,1,0,0,APARTMENT


In [4]:
df.info()
df.isnull().sum()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 25731 entries, 0 to 25730
Data columns (total 19 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   price                        25731 non-null  int64  
 1   bedrooms                     25731 non-null  float64
 2   bathrooms                    25731 non-null  float64
 3   livingArea                   25731 non-null  float64
 4   lotSize                      25731 non-null  float64
 5   yearBuilt                    25731 non-null  float64
 6   propertyTaxRate              25731 non-null  float64
 7   taxAssessedValue             25731 non-null  float64
 8   zestimate                    25731 non-null  float64
 9   rentZestimate                25731 non-null  float64
 10  latitude                     25731 non-null  float64
 11  longitude                    25731 non-null  float64
 12  resoFacts/stories            25731 non-null  float64
 13  resoFacts/hasGar

price                          0
bedrooms                       0
bathrooms                      0
livingArea                     0
lotSize                        0
yearBuilt                      0
propertyTaxRate                0
taxAssessedValue               0
zestimate                      0
rentZestimate                  0
latitude                       0
longitude                      0
resoFacts/stories              0
resoFacts/hasGarage            0
resoFacts/hasHeating           0
resoFacts/hasCooling           0
resoFacts/hasFireplace         0
resoFacts/isNewConstruction    0
homeType                       0
dtype: int64

## One-Hot Encoding

> **Note:** One-Hot Encoding is used to convert categorical data into numerical form.  
> Machine learning models cannot understand text values directly, so categories are transformed into binary columns.  
> This helps the model process and learn patterns more effectively.

In [5]:
df = pd.get_dummies(df, drop_first=True)

In [13]:
df.head()

,price,bedrooms,bathrooms,livingArea,lotSize,yearBuilt,propertyTaxRate,taxAssessedValue,zestimate,rentZestimate,...,resoFacts/hasHeating,resoFacts/hasCooling,resoFacts/hasFireplace,resoFacts/isNewConstruction,homeType_CONDO,homeType_HOME_TYPE_UNKNOWN,homeType_SINGLE_FAMILY,homeType_TOWNHOUSE,house_age,price_per_sqft
0,315400,3.0,3.0,1806.0,1555.0,2007.0,1.12,273960.0,317406.0,1950.0,...,1,0,1,0,False,False,False,True,18.0,174.640089
1,400000,3.0,3.0,1518.0,3484.0,2001.0,1.12,426330.0,573700.0,2545.0,...,1,1,1,0,False,False,True,False,24.0,263.504611
2,512000,4.0,3.0,2724.0,9583.0,1982.0,1.12,397560.0,512355.0,1995.0,...,1,1,1,0,False,False,True,False,43.0,187.958884
3,348000,3.0,1.0,1217.0,13939.0,1953.0,1.12,269520.0,573700.0,1695.0,...,1,0,1,0,False,False,True,False,72.0,285.949055
4,510000,6.0,3.0,2400.0,8545.0,1967.0,1.12,421980.0,573700.0,1395.0,...,1,1,0,0,False,False,False,False,58.0,212.500000


In [18]:
bool_cols = df.select_dtypes(include='bool').columns

df[bool_cols] = df[bool_cols].astype(int)

In [26]:
df.head()

,price,bedrooms,bathrooms,livingArea,lotSize,yearBuilt,propertyTaxRate,taxAssessedValue,zestimate,rentZestimate,...,resoFacts/hasHeating,resoFacts/hasCooling,resoFacts/hasFireplace,resoFacts/isNewConstruction,homeType_CONDO,homeType_HOME_TYPE_UNKNOWN,homeType_SINGLE_FAMILY,homeType_TOWNHOUSE,house_age,price_per_sqft
0,315400,3.0,3.0,1806.0,1555.0,2007.0,1.12,273960.0,317406.0,1950.0,...,1,0,1,0,0,0,0,1,18.0,174.640089
1,400000,3.0,3.0,1518.0,3484.0,2001.0,1.12,426330.0,573700.0,2545.0,...,1,1,1,0,0,0,1,0,24.0,263.504611
2,512000,4.0,3.0,2724.0,9583.0,1982.0,1.12,397560.0,512355.0,1995.0,...,1,1,1,0,0,0,1,0,43.0,187.958884
3,348000,3.0,1.0,1217.0,13939.0,1953.0,1.12,269520.0,573700.0,1695.0,...,1,0,1,0,0,0,1,0,72.0,285.949055
4,510000,6.0,3.0,2400.0,8545.0,1967.0,1.12,421980.0,573700.0,1395.0,...,1,1,0,0,0,0,0,0,58.0,212.500000


## Outlier Removal  


> Outliers are extreme values that can negatively affect model performance.  
> They can distort patterns and reduce prediction accuracy.  
> Removing outliers helps the model learn more reliable and realistic relationships.

In [21]:
df = df[df['price'] < df['price'].quantile(0.95)]

## Feature Engineering  


> Feature engineering involves creating new features from existing data.  
> These new features help the model understand relationships better.  
> It improves prediction accuracy and overall model performance.

In [30]:
df['house_age'] = 2025 - df['yearBuilt']
df['price_per_sqft'] = df['price'] / df['livingArea']

In [37]:
#removing infinity values and filling missing values with median

df.replace([np.inf, -np.inf], np.nan, inplace=True)
df.fillna(df.median(), inplace=True)

## Split Data


> The dataset is split into training and testing sets to evaluate model performance.  
> The model is trained on the training data and tested on unseen data.  
> This helps in checking how well the model generalizes to new data.

In [38]:
X = df.drop('price', axis=1)
y = df['price']

X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42
)

## Linear Regression


>Linear Regression is a fundamental machine learning algorithm used to model the relationship between a dependent variable (house price) and independent variables (features like bedrooms, living area, etc.). 

> Since it is simple and easy to interpret, it helps in identifying how different features influence the target variable.  
> It also acts as a baseline model, allowing us to compare and measure improvements when using more advanced models.

In [39]:
lr = LinearRegression()
lr.fit(X_train, y_train)

y_pred_lr = lr.predict(X_test)
print("Linear R2:", r2_score(y_test, y_pred_lr))

Linear R2: 0.8526837150278067


In [40]:
# Train
y_train_pred_lr = lr.predict(X_train)
print("Linear Train R2:", r2_score(y_train, y_train_pred_lr))

# Test
y_test_pred_lr = lr.predict(X_test)
print("Linear Test R2:", r2_score(y_test, y_test_pred_lr))

Linear Train R2: 0.8904121709547578
Linear Test R2: 0.8526837150278067


In [41]:
# MSE
mse_lr = mean_squared_error(y_test, y_pred_lr)

# RMSE
rmse_lr = np.sqrt(mse_lr)

print("Linear R2:", r2_score(y_test, y_pred_lr))
print("Linear MSE:", mse_lr)
print("Linear RMSE:", rmse_lr)

Linear R2: 0.8526837150278067
Linear MSE: 4020310842.429827
Linear RMSE: 63405.92119376413


## Key Metrics

> **R² Score on Training Data (Train R²)**: This is about 0.90, which means the model explains 90% of the variation in house prices using the training data. It's a decent fit to what it learned from.

> **R² Score on Test Data (Test R²)**: This is also around 0.90, so it performs similarly well on new data it hasn't seen before. No big drop, which is good.

> **Mean Squared Error (MSE)**: This is about 4,020,310,842, measuring the average squared difference between predicted and actual prices. It's a large number, but lower than some other models.

> **Root Mean Squared Error (RMSE)**: About $63,406, which is the square root of MSE. Think of it as the average error in dollars – on average, our predictions are off by around $63,406. It's okay, but not the best.

Overall, Linear Regression does a solid job as a starting point, but it might miss some complex patterns in house prices. It's simple and reliable, but we can do better with more advanced models.

## Polynomial Regression

> Polynomial regression is like a fancy version of linear regression. Instead of just drawing a straight line to fit the data, it uses curves by adding powers of the features (like x squared, x cubed, etc.). This helps capture more complex patterns that a straight line can't handle.

> We're using it here because house prices might not follow a simple straight-line relationship. By going polynomial, we can model those twists and turns in the data better, hopefully getting more accurate predictions than plain linear regression.

In [42]:
poly = PolynomialFeatures(degree=2)

X_train_poly = poly.fit_transform(X_train)
X_test_poly = poly.transform(X_test)

model_poly = LinearRegression()
model_poly.fit(X_train_poly, y_train)

y_pred_poly = model_poly.predict(X_test_poly)
print("Polynomial R2:", r2_score(y_test, y_pred_poly))

Polynomial R2: 0.999727717533278


In [43]:
# Train
y_train_pred_poly = model_poly.predict(X_train_poly)
print("Poly Train R2:", r2_score(y_train, y_train_pred_poly))

# Test
y_test_pred_poly = model_poly.predict(X_test_poly)
print("Poly Test R2:", r2_score(y_test, y_test_pred_poly))

Poly Train R2: 0.9998299402842945
Poly Test R2: 0.999727717533278


## Key Metrics for Polynomial Regression

> **R² Score on Training Data (Train R²)**: This is about 1.00, which means the model explains almost all of the variation in house prices using the training data. It's very good at fitting what it learned from.

> **R² Score on Test Data (Test R²)**: This is also around 1.00, so it performs just as well on new data it hasn't seen before. No signs of overfitting.

> **Mean Squared Error (MSE)**: This is a small number like 70,722, measuring the average squared difference between predicted and actual prices. Lower is better, and this is much lower than linear regression.

> **Root Mean Squared Error (RMSE)**: About 266, which is the square root of MSE. Think of it as the average error in dollars – on average, our predictions are off by around $266. That's a big improvement over linear regression.

Overall, Polynomial Regression does an excellent job, almost perfectly predicting house prices, but we should watch for overfitting since it's so close to perfect.



## Random Forest

> Random Forest is like a team of decision trees working together to make smarter predictions. Each tree looks at the data and makes its own guess about the house price based on features like bedrooms, location, and age. Then, the forest combines all these guesses (usually by averaging them) to give a final, more accurate answer.

> It's useful here because house prices depend on many interconnected factors, and Random Forest can handle that complexity without overfitting too much. It also works well with both numerical and categorical data, making it a strong choice for real estate prediction. Compared to simpler models like Linear Regression, it often captures non-linear patterns better, leading to higher accuracy on unseen data.


In [44]:
rf = RandomForestRegressor(
    n_estimators=300,
    max_depth=15,
    random_state=42
)

rf.fit(X_train, y_train)

y_pred_rf = rf.predict(X_test)
print("Random Forest R2:", r2_score(y_test, y_pred_rf))

Random Forest R2: 0.9827941452721763


In [45]:
# Train
y_train_pred_rf = rf.predict(X_train)
print("RF Train R2:", r2_score(y_train, y_train_pred_rf))

# Test
y_test_pred_rf = rf.predict(X_test)
print("RF Test R2:", r2_score(y_test, y_test_pred_rf))

RF Train R2: 0.9962498559242737
RF Test R2: 0.9827941452721763


In [46]:
mse_rf = mean_squared_error(y_test, y_pred_rf)
rmse_rf = np.sqrt(mse_rf)

print("RF R2:", r2_score(y_test, y_pred_rf))
print("RF MSE:", mse_rf)
print("RF RMSE:", rmse_rf)

RF R2: 0.9827941452721763
RF MSE: 469553548.1946113
RF RMSE: 21669.1842992442


## Key Metrics for Random Forest

> **R² Score on Training Data (Train R²)**: This is about 0.996, which means the model explains 99.6% of the variation in house prices on the data it learned from. It fits the training data extremely well.

> **R² Score on Test Data (Test R²)**: This is around 0.979, so it also performs very well on new data it hasn't seen before. That means it generalizes nicely and does not overfit too much.

> **Mean Squared Error (MSE)**: This is about 469,553,548, measuring the average squared difference between predicted and actual prices. Smaller is better, and this is much lower than the linear regression model.

> **Root Mean Squared Error (RMSE)**: About $21,669, which is the average dollar error in predictions. On average, the model is off by roughly $21,700.

Overall, Random Forest gives strong performance with high accuracy on both training and test data, making it a reliable choice for predicting house prices.

## Gradient Boosting

> Gradient Boosting is like a smart team of decision trees that work together to make better predictions. It starts with one simple tree, then adds more trees one by one, each fixing the mistakes of the previous ones. This way, it learns from errors and gets more accurate over time.

> It's useful here for predicting house prices because real estate data is complex with many factors (like location, size, and age) that don't follow simple patterns. Gradient Boosting handles this well, often giving better results than basic models, and it can spot important features to improve predictions. Compared to Random Forest, it focuses on correcting errors, leading to strong performance on both training and new data.

In [47]:
gb = GradientBoostingRegressor()

gb.fit(X_train, y_train)

y_pred_gb = gb.predict(X_test)
print("Gradient Boosting R2:", r2_score(y_test, y_pred_gb))

Gradient Boosting R2: 0.9747054673713487


In [48]:
# Train
y_train_pred_gb = gb.predict(X_train)
print("GB Train R2:", r2_score(y_train, y_train_pred_gb))

# Test
y_test_pred_gb = gb.predict(X_test)
print("GB Test R2:", r2_score(y_test, y_test_pred_gb))

GB Train R2: 0.9795719772988616
GB Test R2: 0.9747054673713487


In [49]:
mse_gb = mean_squared_error(y_test, y_pred_gb)
rmse_gb = np.sqrt(mse_gb)

print("GB R2:", r2_score(y_test, y_pred_gb))
print("GB MSE:", mse_gb)
print("GB RMSE:", rmse_gb)

GB R2: 0.9747054673713487
GB MSE: 690296281.9104265
GB RMSE: 26273.490097633137


## Key Metrics for Gradient Boosting

> **R² Score on Training Data (Train R²)**: This is about 0.982, which means the model explains 98.2% of the variation in house prices on the training data. It's an excellent fit to what it learned from.

> **R² Score on Test Data (Test R²)**: This is around 0.974, so it performs very well on new data it hasn't seen before. The small drop from training to test shows the model generalizes nicely without significant overfitting.

> **Mean Squared Error (MSE)**: This is about 690,296,282, measuring the average squared difference between predicted and actual prices. It's lower than linear regression but higher than random forest.

> **Root Mean Squared Error (RMSE)**: About $26,273, which is the average error in dollars. On average, predictions are off by around $26,300, which is better than linear regression but slightly higher than random forest.

Overall, Gradient Boosting delivers strong and reliable performance, with good accuracy on both training and test data. It's a solid choice for real-world house price predictions, balancing accuracy with generalization well.

## Overall Report

- **Data loading and preprocessing**
    - Read `cleaned_housing.csv`
    - One-hot encoded categorical home type columns
    - Converted boolean columns to integers
    - Removed top 5% of price outliers
    - Engineered `house_age` and `price_per_sqft`
    - Replaced infinity values and filled missing values with median

- **Train/test split**
    - Features: all columns except `price`
    - Target: `price`
    - Split into 80% training and 20% testing

- **Model performance**
    - **Linear Regression**
        - Train R² ≈ 0.90
        - Test R² ≈ 0.90
        - RMSE ≈ $63,406
        - Good baseline, but relatively high error for house price prediction

    - **Polynomial Regression (degree 2)**
        - Train R² ≈ 1.00
        - Test R² ≈ 1.00
        - RMSE ≈ $266
        - Extremely high fit; likely too complex and may overfit despite excellent metrics

    - **Random Forest**
        - Train R² ≈ 0.996
        - Test R² ≈ 0.979
        - RMSE ≈ $21,669
        - Best generalization among models, strong accuracy on unseen data

    - **Gradient Boosting**
        - Train R² ≈ 0.982
        - Test R² ≈ 0.974
        - RMSE ≈ $26,273
        - Strong performance, slightly behind Random Forest

- **Conclusion**
    - The dataset is well-prepared and the models are performing strongly.
    - Random Forest is the best choice for this task based on test R² and RMSE.
    - Gradient Boosting is also a strong alternative.
    - Linear Regression is useful as a baseline.
    - Polynomial Regression may show overfitting; validate further with cross-validation before trusting it fully.